joomla/joomla-cms
# Allow URLs with lengths up to 2048 chars in com_redirect, redo of PR #4781

Redo of Pull Request #4781 for 3.5.0.
#### Summary of Changes

This is a redo of PR #4781 to allow longer URLs in com_redirect, thanks to @Kixo for his work.

**This PR is not only an improvement, it also fixes problems with not unique URLs after conversion from utf8 to utf8mb4 and so should go into 3.5.0 RC.**

All 3 URL columns for old and new URL and the referrer are enlarged to max. 2048 characters.

I had to change a few things compared to the original #4781:
- Use 2048 = sitemap protocol limit and not 2083 = old IE limit
- Use new update sql file names
- Adapt to latest changes for utf8mb4
- Not completely remove the index but change it from a unique key to just a normal index, and on mysql limit the index to the 1st 100 chars for utf8mb4 compatibility.

What this PR does **not** change is that if the backend user creates a redirect link manually, i.e. by creating a new one in the backend, he is not warned or alerted when entering a too long URL.

**Beside** the enlargement of the URL colums, **this PR corrects 2 older update sql scripts**, where "KEY" insted of "INDEX" was used for adding or removing indexes, which is valid syntax but not recognized by the schema manager when testing schema status. This had no visible and testable effect when I tested, so maybe has to be accepted by code review.
#### Testing Instructions

This PR has to be tested with **both new installations** of current staging + this patch, find the source here: [https://github.com/richard67/joomla-cms/archive/long-urls-in-redirect-component.zip](https://github.com/richard67/joomla-cms/archive/long-urls-in-redirect-component.zip)

**and** with **updating** from a 3.4.8 to a current staging + this patch, find the source for updating with the extension installer here:  [http://test5.richard-fath.de/Joomla_3.5.0-beta3-Beta-Update_Package_test1.zip](http://test5.richard-fath.de/Joomla_3.5.0-beta3-Beta-Update_Package_test1.zip)

or for updating with Joomla! update component using this as custom URL, here: [http://test5.richard-fath.de/list_test1.xml](http://test5.richard-fath.de/list_test1.xml)

or instead of updating, apply this patch e.g. with the patch tester component to a current staging.

After an update (or apply pacth) you have to go to "Extensions -> Manage -> Database".

There should be no database problems reported about redirect_links table.

Fix database problems (at least the utf8mb4 conversion should be shown to be done) until all done. 

Then on both new installation and updated system play around with
- automatically creating redirect links by calling a not existing URL with length of 2048 characters in your domain, or
- manually creating redirect links with old URLs having a lenght of 2048 characters in your domain

Test all aspects, e.g. enabling the redirect to an existing URL in your domain, sorting by old URL if you have created a few ones and so on.

Result: All should work as well as before but now with the long URLs.

A long URL to redirect to will not be easy to test because we cannot make just 1 menu item with (2048 - lenght of "protocol + domain name") characters long alias to have a long URL where we could redirect too :smile:

**Note**: If we wanna test every aspect we **need testers for Postgresql and Microsoft SQL (sqlazure), too**.

One of the links regarding URL lengths which made me chose the 2048 as limit:

[http://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers](http://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers)
